Get rid of fb_utilities

bhalevy opened the pull request:
utils::fb_utilities is a global in-memory registry for storing and retrieving broadcast_address and broadcat_rpc_address.
As part of the effort to get rid of all global state, this series gets rid of fb_utilities.
This will eventually allow e.g. cql_test_env to instantiate multiple scylla server nodes, each serving on its own address.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5003/)
:white_check_mark: - [Sanity Tests](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6361/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5003/)
#### Build Details:
- Duration: 2 hr 24 min
- Builder: spider2.cloudius-systems.com

avikivity commented:
Very good, it was a bad wart.

bhalevy commented on locator/topology.hh:
```diff
@@ -336,6 +337,28 @@ public:
 
+    host_id my_host_id() const noexcept {
+        auto* node = this_node();
+        return node ? node->host_id() : _cfg.this_host_id;
```
This_node gets its host_id initially from `_cfg`.
Theoretically it is possible to change it later on, but we don't and the node is added only after calling `set_host_id_cfg`.
So it should be safe to always return `_cfg.this_host_id`.

bhalevy commented:
In [9314d99](https://github.com/scylladb/scylladb/commit/9314d99f718b0a0c0fad57a47374ac59be07332f):
- rebased
- locator: topology: add helpers to retrieve this host_id and address
  - simplified. always return `_cfg` values.
- gossiper: add get_this_endpoint_state_ptr
  - refactored out of next patch
  - renamed method to `get_this_endpoint_state_ptr`

xemul commented on message/messaging_service.hh:
```diff
@@ -342,6 +343,9 @@ public:
     uint16_t port();
     gms::inet_address listen_address();
+    gms::inet_address broadcast_address() {
+        return _cfg.broadcast_address;
```
Messaging service already points to token metadata, as it also needs topology. It could retrieve broadcast address from there as well

bhalevy replied:
Nope, apparently we can't rely on _token_metadata early enough, e.g. from:
```
netw::messaging_service::broadcast_address() const at ./message/messaging_service.cc:487
service::retrying_dispatcher::dispatch_to_node(netw::msg_addr, query::forward_request) at ./service/forward_service.cc:270
operator() at ./service/forward_service.cc:602
```

As hit by `topology.test_change_ip`

xemul replied:
Pity. Not this time then

xemul reviewed (approved):
LGTM except for some minor nits that can be postponed

bhalevy commented:
In [31a66f2](https://github.com/scylladb/scylladb/commit/31a66f225c42ab6fd5cb4290e50aec855124482b):
- db/system_keyspace: save_local_info: get broadcast addresses from caller
- db/system_keyspace: use topology via db rather than fb_utilities
- database: use locator::topology rather than fb_utilities
  - do not expose get_topology in database.hh
- messaging_service: move listen_address and port getters inline
- messaging_service: accept broadcast_addr in config rather than via fb_utilities
  - Added commentary for `config.{ip,broadcast_address}` (did not rename `ip` member to `listen_address`.  can be done in followup)

bhalevy commented:
In v4 ([0bcce35](https://github.com/scylladb/scylladb/commit/0bcce35abd4c8b78941dee37e4c965eae3889220)):
- rebased
- messaging_service: move listen_address and port getters inline
- messaging_service: accept broadcast_addr in config rather than via fb_utilities
- got rid of database::get_topology (callers need to go through db.get_token_metadata to get to the topology)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5068/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6430/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5068/)
#### Build Details:
- Duration: 2 hr 31 min
- Builder: spider7.cloudius-systems.com
